File: 0064-minimum-path-sum/0064-minimum-path-sum.swift
# @param {Integer[][]} grid
# @return {Integer}
def min_path_sum(grid)
  grid.each_index do |i|
    grid.first.each_index do |j|
      if i == 0 && j == 0
        grid[i][j] = grid[i][j]
      elsif i == 0 && j > 0
        grid[i][j] = grid[i][j] + grid[i][j-1]
      elsif j == 0 && i > 0
        grid[i][j] = grid[i][j] + grid[i-1][j]
      else
        grid[i][j] = grid[i][j] + [grid[i][j-1], grid[i-1][j]].min
      end
    end
  end
  grid[-1][-1]
end

File: 0064-minimum-path-sum/0064-minimum-path-sum.scala
object Solution {
  def minPathSum(grid: Array[Array[Int]]): Int = {
    (grid.length, grid.headOption.map(_.length).getOrElse(0)) match {
      case (m, n) if m > 0 && n > 0 =>
        val dp = Array.fill(m, n)(grid(m - 1)(n - 1))
        for {
          k <- 1 to m + n
          i <- math.max(0, (m - 1) - k) to math.min(m - 1, m + n - k - 2)
        } {
          val j = m + n - i - k - 2
          dp(i)(j) = grid(i)(j) + Seq((i + 1, j), (i, j + 1))
            .collect {
              case (x, y) if dp.isDefinedAt(x) && dp(x).isDefinedAt(y) => dp(x)(y)
            }
            .min
        }
        dp(0)(0)
    }
  }
}